hydradx/model/amm/omnipool_amm.py

```diff
@@ -1,36 +1,71 @@
-from numbers import Number
 from typing import Callable
+import math
 
 from .agents import Agent
-from .amm import AMM, FeeMechanism, basic_fee
+from .exchange import Exchange
 from .oracle import Oracle, Block, OracleArchiveState
 from .stableswap_amm import StableSwapPoolState
+from typing import Literal
 
+class DynamicFee:
 
-class OmnipoolState(AMM):
+    def __init__(
+        self,
+        minimum: float = 0,
+        maximum: float = float('inf'),
+        amplification: float = 0,
+        decay: float = 0,
+        # ^^ if these four are provided, we can figure the rest out.
+        current: dict[str: float] = None,
+        liquidity: dict = None,
+        net_volume: dict = None,
+        last_updated: dict = None
+    ):
+        self.amplification = amplification
+        self.decay = decay
+        self.minimum = minimum
+        self.maximum = maximum
+        if current is None:
+            self.current = {}
+        else:
+            self.current = current
+        self.last_updated = last_updated or {tkn: 0 for tkn in self.current}
+        self.last_liquidity = liquidity if liquidity is not None else {}
+        self.last_volume = net_volume if net_volume is not None else {}
+
+    def update(self, time_step: int, volume: dict, liquidity: dict):
+        for tkn in self.current:
+            if self.last_updated[tkn] == time_step:
+                # update only when fee[tkn] has been accessed this block
+                self.last_liquidity[tkn] = liquidity[tkn]
+                self.last_volume[tkn] = volume[tkn]
+
+
+class OmnipoolState(Exchange):
     unique_id: str = 'omnipool'
+    _asset_fee: DynamicFee
+    _lrna_fee: DynamicFee
 
     def __init__(self,
                  tokens: dict[str: dict],
                  tvl_cap: float = float('inf'),
                  preferred_stablecoin: str = None,
-                 asset_fee: dict or FeeMechanism or float = None,
-                 lrna_fee: dict or FeeMechanism or float = None,
+                 asset_fee: dict or DynamicFee or float = 0,
+                 lrna_fee: dict or DynamicFee or float = 0,
                  oracles: dict[str: int] = None,
                  trade_limit_per_block: float = float('inf'),
                  update_function: Callable = None,
-                 last_asset_fee: dict or float = None,
-                 last_lrna_fee: dict or float = None,
-                 imbalance: float = 0.0,
                  last_oracle_values: dict = None,
                  max_withdrawal_per_block: float = 1,
                  max_lp_per_block: float = float('inf'),
                  remove_liquidity_volatility_threshold: float = 0,
                  withdrawal_fee: bool = True,
                  min_withdrawal_fee: float = 0.0001,
-                 lrna_mint_pct: float = 0.0,
-                 unique_id: str = 'omnipool'
+                 lrna_mint_pct: float = 1.0,
+                 unique_id: str = 'omnipool',
+                 lrna_fee_burn: float = 1.0,
```

Comment:
Default lrna_fee_burn should be 0.5, since that's what we're doing in PROD. Eventually though we should replace this with parameter sets.

Reply:
done